### ISSUE TITLE
some issues with the side bar

### ISSUE THREAD

**Issue by vigor-vlad** [gridstack/gridstack.js#667]
Hi there. First of all, thanks for your great library. However I've encountered a number of issues. I'm using it in an angular app. The app itself is a dashboard and a sidebar, the user should be able to drag'n'drop widgets between them. Here are the issues that I've encountered (jsfiddle: [https://jsfiddle.net/pyakphaa/9/](url) ):
1. It's hard to drop the widget from the dashboard to the sidebar. The item is dropped only if positioned around 2/5 of the width of the sidebar:
![issue-1](https://cloud.githubusercontent.com/assets/24697907/26156146/1c2133d4-3b1e-11e7-9432-58c81ff90880.png)
I was able to bypass this issue by adding a "inTrashZone":
```javascript
                this.dd.on(trashZone, 'dropover', function(event, ui) {
                    /*...*/
                    el.data('inTrashZone', true);
                    /*...*/
                }).on(trashZone, 'dropout', function(event, ui) {
                    /*...*/
                    el.data('inTrashZone', false);
                    /*...*/
                });`
```
 I also had to add an extra condition to the "if (event.type == 'drag') { if (el.data('inTrashZone') || x < 0 || x >= self.grid.width || y < 0 || (!self.grid.float && y > self.grid.getGridHeight())) {" (in "dragOrResize" function)
But I'm not sure this is a good approach...

2. When the widget is dropped there is an exception:
`VM812:241 Uncaught TypeError: Cannot read property 'el' of undefined
    at Object.vm.onItemRemoved (https://fiddle.jshell.net/_display/:241:28)
    at fn (eval at compile (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.9/angular.min.js:236:126), <anonymous>:4:307)
    at m.d.(anonymous function) [as onItemRemoved] (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.9/angular.min.js:84:300)
    at HTMLDivElement.<anonymous> (https://kdietrich.github.io/gridstack-angular/dist/gridstack-angular.min.js:10:1200)
    at HTMLDivElement.dispatch (https://code.jquery.com/jquery-3.2.1.min.js:3:10316)
    at HTMLDivElement.q.handle (https://code.jquery.com/jquery-3.2.1.min.js:3:8343)
    at Object.trigger (https://code.jquery.com/jquery-3.2.1.min.js:4:5628)
    at r.fn.init.triggerHandler (https://code.jquery.com/jquery-3.2.1.min.js:4:6222)
    at Function.xa.fn.on.xa.cleanData (https://ajax.googleapis.com/ajax/libs/angularjs/1.5.9/angular.min.js:23:70)
    at Ka (https://code.jquery.com/jquery-3.2.1.min.js:3:15966)
    at r.fn.init.remove (https://code.jquery.com/jquery-3.2.1.min.js:3:16852)`
Looks like there is a conflict between the gridstack library and its angular extension. I was able to fix it by removing the "el.removeData('_gridstack_node');" line in "if (node._isAboutToRemove) {" ("onEndMoving" function)

3. Another issue is caused by overlapping grid and sidebar. To reproduce it you'll have to do the following: d'n'd the same widget to the sidebar, then to the grid, back to the sidebar and once again to the grid. When you d'n'd the widget from the sidebar to the grid for the second time the "dropover" event of the grid is not triggered somehow, the placeholder is not shown and if you drop the widget to the grid it will crash. But if you drag the widget outside the grid the "dropout" event is fired and then you can drag it back into the grid and it will work OK. 

4. And the last issue: When you drag the widget from the sidebar to the grid you are unable to place it in the last row if there are widgets in the grid already. You can place it only above the existing widgets.

Hope you can help me with those issues. Please let me know if you need any more details.

Best regards,
Vlad

**Comment by vigor-vlad** [comment]
a small update on the first issue. There we still some problems on dropping the item into the trash zone, for example when the width of the item is greater then the width of the  trash zone. And in "gridstack.jQueryUI.js" I found some strange code:
```javascript
    JQueryUIGridStackDragDropPlugin.prototype.droppable = function(el, opts) {
        el = $(el);
        if (opts === 'disable' || opts === 'enable') {
            el.droppable(opts);
        } else {
            el.droppable({
                accept: opts.accept
            });
        }
        return this;
    };
```
It prevents you from using advanced droppable options, like 'tolerance', 'greedy', etc. Here is how I fixed it:
```javascript
    JQueryUIGridStackDragDropPlugin.prototype.droppable = function(el, opts) {
        el = $(el);
        el.droppable(opts);
        return this;
    };
```
I have the question. Did you make this restriction on purpose?

So, the overall solution for the 'Side bar'/'Trash zone' droppable looks like:

1. In "gridstack.jQueryUI.js" replace
```javascript
    JQueryUIGridStackDragDropPlugin.prototype.droppable = function(el, opts) {
        el = $(el);
        if (opts === 'disable' || opts === 'enable') {
            el.droppable(opts);
        } else {
            el.droppable({
                accept: opts.accept
            });
        }
        return this;
    };
```
with
```javascript
    JQueryUIGridStackDragDropPlugin.prototype.droppable = function(el, opts) {
        el = $(el);
        el.droppable(opts);
        return this;
    };
```
2. In gridstack.js replace
```javascript
            if (!this.dd.isDroppable(trashZone)) {
                this.dd.droppable(trashZone, {
                    accept: '.' + self.opts.itemClass
                });
            }
```
with
```javascript
            if (!this.dd.isDroppable(trashZone)) {
                this.dd.droppable(trashZone, {
                    tolerance : 'touch',
                    greedy : true,
                    accept: '.' + self.opts.itemClass,
                    hoverClass : self.opts.removableHoverClass
                });
            }
```
*these 'trashZone' options should probably be moved to the gridstack options...

3. add 'inTrashZone' flag and set it and drop it when needed:
```javascript
            this.dd
                .on(trashZone, 'dropover', function(event, ui) {
                    //...
                    el.data('inTrashZone', true);
                    //...
                })
                .on(trashZone, 'dropout', function(event, ui) {
                    //...
                    el.data('inTrashZone', false);
                    //...
                });
```

4. check this 'inTrashZone' flag in:
```javascript
        var dragOrResize = function(event, ui) {
            //...
            if (event.type == 'drag') {
                if (el.data('inTrashZone') /* || ...*/) {
                    //...
                }
               //...
            }
            //...
        };
```

P.S. I prefer not to mess with the libraries written by other people, but I wasn't able to find any other solution. Hope it's OK 8)

**Comment by radiolips** [comment]
Hi, @vigor-vlad ! I believe that the code is not intentionally exclusive; adding more user-defined options to `droppable` is not a problem. We may want to add different option parameters when defining the grid for trash, sidebar, and widgets. Feel free to add a PR if you'd like! Otherwise, I'll take a look at what you posted and some of the code and see if I can't get `droppable` options passed through!

**Comment by radiolips** [comment]
At least some of this was resolved in `v0.4.0`. I'm closing this - let me know if you have further problems.
